specifications/locks_auxiliary_vars/LockHS.tla

```diff
@@ -0,0 +1,360 @@
+-------------------------------- MODULE LockHS --------------------------------
+
+(*****************************************************************************)
+(* This module contains the specification of a lock with auxiliary variables.*)
+(* 1. A history variable `h_turn` is needed to remember the assignement of   *)
+(*    the turn variable used inside the Peterson specification.              *)
+(* 2. A stuttering variable `s` is added to force the stuttering of the Lock *)
+(*    specification to mimick the 3 steps taken by Peterson to enter the     *)
+(*    critical section.                                                      *)
+(* With these variables, one can finally refine LockHS to Peterson, giving   *)
+(* an equivalence between the Lock and Peterson specifications.              *)
+(*                                                                           *)
+(* The stuttering is achieved using the Stuttering module created by Leslie  *)
+(* Lamport and comes from to the paper "Auxiliary Variables in TLA+".        *)
+(* The module used here has been modified, see explanations at the end of    *) 
+(* the Stuttering module.                                                    *)
+(*****************************************************************************)
+
+EXTENDS Lock
+
+\* History variable to remember the turn variable
+VARIABLE h_turn
+NoHistoryChange(A) == A /\ UNCHANGED h_turn
+
+\* Stuttering variable
+VARIABLE s
+INSTANCE Stuttering
+
+ASSUME AltStutterConstantCondition(1..2, 1, LAMBDA j : j-1)
```

Comment:
Why is this an assumption? The predicate boils down to
``ReachBot[{1}] = 1 .. 2``
(with ReachBot defined in the definition of the predicate) and indeed a quick calculation shows that
``ReachBot[{1}] = ReachBot[{1,2}] = {1,2}``
so the condition holds, so this should really be a theorem. The TLAPS proof of that theorem may be tricky because it involves a recursive function (which, moreover, is defined inside a ``LET``), if you have trouble with it, I can have a try at it.